### ISSUE TITLE
Garbage Collection of Formulas for Solvers written in C

### ISSUE THREAD

**Issue by PhilippWendler** [sosy-lab/java-smt#1]
For solvers written in C, we need to manually free objects like formulas. We can do so automatically after the corresponding Java object has been freed by using `PhantomReference`. This is implemented for Z3 and can be ported to other solvers as well, or maybe we can even write a common implementation. It should be possible to disable this feature such that users that do not need it do not need to pay the cost of PhantomReferences.

**Comment by cheshire** [comment]
Would also make for a nice comparison on finalizers vs. phantom references vs. leaking memory.

**Comment by cheshire** [comment]
We should also garbage collect contexts (if the client decides to make bazillion `FormulaManager`s).
That is blocked by #6, as we currently do not have a single instance which would be aware of all created `FormulaManager`s

**Comment by PhilippWendler** [comment]
`FormulaManager`s where freeing them is relevant have an explicit `close` method and should be closed manually.

**Comment by cheshire** [comment]
For Z3 we should compare PhantomReference's (our implementation) vs. finalizers (Z3 own implementation) vs. leaking memory.

Mathsat5 exposes garbage-collection interface: 

```c
/**
 * \brief Performs garbage collection on the given environment
 *
 * This function will perform garbage collection on the given environment.
 * All the internal caches of the environment will be cleared (including those
 * in the active solvers and preprocessors). If the environment is not shared,
 * all the terms that are not either in \a tokeep or in the current asserted
 * formulas will be deleted.
 *
 * \param env The environment in which to operate.
 * \param tokeep List of terms to not delete.
 * \param num_tokeep Size of the \a tokeep array.
 * \return zero on success, nonzero on error.
 */
int msat_gc_env(msat_env env, msat_term *tokeep, size_t num_tokeep);
```

we should expose calling it (as an option, perhaps every N allocations?).
An array of relevant `msat_term`s can be stored in `Msat5FormulaCreator` by storing `WeakReference`s to them.

**Comment by cheshire** [comment]
Closing in favor of #39
